File: 01_Kafka_Clients/java/src/main/java/com/masterclass/kafka/OrderProducer.java
package com.masterclass.kafka;

import io.confluent.kafka.serializers.KafkaAvroSerializer;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;

import java.time.Instant;
import java.util.List;
import java.util.Properties;
import java.util.UUID;

public final class OrderProducer {
    private static final List<String> STATUSES = List.of("PLACED", "PAID", "SHIPPED");

    public static void main(String[] args) throws Exception {
        int orderCount = args.length > 0 ? Integer.parseInt(args[0]) : 3;

        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, KafkaConfig.BOOTSTRAP_SERVERS);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, KafkaAvroSerializer.class.getName());
        props.put("schema.registry.url", KafkaConfig.SCHEMA_REGISTRY_URL);
        props.put(ProducerConfig.ACKS_CONFIG, "all");
        props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);

        try (KafkaProducer<String, OrderEvent> producer = new KafkaProducer<>(props)) {
            System.out.printf("Publishing %d order(s) to topic '%s' ...%n", orderCount, KafkaConfig.TOPIC);

            for (int i = 0; i < orderCount; i++) {
                String orderId = "ord-" + UUID.randomUUID().toString().substring(0, 8);
                String customerId = "cust-" + (1000 + i);
                long amountCents = 4999L + (i * 1500L);

                for (String status : STATUSES) {
                    OrderEvent event = OrderEvent.newBuilder()
                            .setOrderId(orderId)
                            .setCustomerId(customerId)
                            .setStatus(status)
                            .setAmountCents(amountCents)
                            .setCurrency("USD")
                            .setEventTime(Instant.now().toEpochMilli())
                            .build();

                    ProducerRecord<String, OrderEvent> record =
                            new ProducerRecord<>(KafkaConfig.TOPIC, orderId, event);

                    producer.send(record, (metadata, exception) -> {
                        if (exception != null) {
                            System.err.printf("Delivery failed: %s%n", exception.getMessage());
                            return;
                        }
                        System.out.printf(
                                "Produced order_id=%s status=%s partition=%d offset=%d%n",
                                orderId, status, metadata.partition(), metadata.offset());
                    }).get();
                    Thread.sleep(100);
                }
            }
        }

        System.out.println("Producer finished.");
    }

    private OrderProducer() {}
}

File: 01_Kafka_Clients/python/order_producer.py
#!/usr/bin/env python3
"""Publish order lifecycle events to Kafka using Avro and Schema Registry."""

from __future__ import annotations

import argparse
import sys
import time
import uuid

from confluent_kafka import SerializingProducer
from confluent_kafka.serialization import StringSerializer
from confluent_kafka.schema_registry import SchemaRegistryClient
from confluent_kafka.schema_registry.avro import AvroSerializer

from config import BOOTSTRAP_SERVERS, SCHEMA_PATH, SCHEMA_REGISTRY_URL, TOPIC

STATUSES = ("PLACED", "PAID", "SHIPPED")


def build_producer() -> SerializingProducer:
    schema_str = SCHEMA_PATH.read_text(encoding="utf-8")
    registry = SchemaRegistryClient({"url": SCHEMA_REGISTRY_URL})
    avro_serializer = AvroSerializer(registry, schema_str)

    return SerializingProducer(
        {
            "bootstrap.servers": BOOTSTRAP_SERVERS,
            "key.serializer": StringSerializer("utf_8"),
            "value.serializer": avro_serializer,
            "acks": "all",
            "enable.idempotence": True,
        }
    )


def delivery_report(err, msg) -> None:
    if err is not None:
        print(f"Delivery failed: {err}", file=sys.stderr)
        return
    key = msg.key().decode("utf-8") if isinstance(msg.key(), bytes) else msg.key()
    print(f"Produced order_id={key} partition={msg.partition()} offset={msg.offset()}")


def publish_orders(producer: SerializingProducer, order_count: int) -> None:
    for i in range(order_count):
        order_id = f"ord-{uuid.uuid4().hex[:8]}"
        customer_id = f"cust-{1000 + i}"
        amount_cents = 4999 + (i * 1500)

        for status in STATUSES:
            event = {
                "order_id": order_id,
                "customer_id": customer_id,
                "status": status,
                "amount_cents": amount_cents,
                "currency": "USD",
                "event_time": int(time.time() * 1000),
            }
            producer.produce(
                topic=TOPIC,
                key=order_id,
                value=event,
                on_delivery=delivery_report,
            )
            print(f"Sent order_id={order_id} status={status}")
            producer.poll(0)
            time.sleep(0.1)


def main() -> int:
    parser = argparse.ArgumentParser(description="Order event producer")
    parser.add_argument("--orders", type=int, default=3, help="Number of orders to simulate")
    args = parser.parse_args()

    producer = build_producer()
    print(f"Publishing {args.orders} order(s) to topic '{TOPIC}' ...")
    publish_orders(producer, args.orders)
    producer.flush()
    print("Producer finished.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
